File: book/python_algorithm_interview/problems/3sum.py
def F(nums):
    nums.sort()
    res = []

    for curr in range(len(nums)-1):
        # 양수/중복스킵
        if (nums[curr] > 0) or (curr > 0 and nums[curr] == nums[curr-1]):
            continue

        left, right = curr+1, len(nums)-1
        while left < right:

            # right 중복스킵
            if right < len(nums)-1 and nums[right] == nums[right+1]:
                right -= 1
                continue

            # left 중복스킵 (하지만 시작 시점에서는 스킵 제외)
            if left > curr+1 and nums[left] == nums[left-1]:
                left += 1
                continue

            # 투포인터
            curSum = nums[curr]+nums[left]+nums[right]
            if curSum == 0:
                res.append([nums[curr], nums[left], nums[right]])
                left += 1
                right -= 1
            elif curSum < 0:
                left += 1
            else:
                right -= 1

    return res


nums = [-5, -5, -4, -4, -2, -2, 0, 0, 0, 1, 1, 1, 3, 4, 4]
print("result: ", F(nums))
# => [[-5, 1, 4], [-4, 0, 4], [-4, 1, 3], [-2, -2, 4], [-2, 1, 1], [0, 0, 0]]

File: achieve/book/python_algorithm_interview/problems/3sum.ts
function sum(nums: number[]) {
  nums.sort();
  let res = [];

  for (let i = 0; i < nums.length - 1; i++) {
    if (nums[i] > 0 || (i > 0 && nums[i] === nums[i - 1])) continue;

    let left = i + 1;
    let right = nums.length - 1;

    while (left < right) {
      if (right < nums.length - 1 && nums[right] === nums[right + 1]) {
        right--;
        continue;
      }

      let sum = nums[i] + nums[left] + nums[right];
      if (sum === 0) {
        res = [...res, [nums[i], nums[left], nums[right]]] as any;
        left++;
        right--;
      } else if (sum < 0) {
        left++;
      } else {
        right--;
      }
    }
  }

  return res;
}

const nums = [-5, -5, -4, -4, -2, -2, 0, 0, 0, 1, 1, 1, 3, 4, 4];
